### PR TITLE
Add % operator and propagate it from parser -> halide -> isl

### LINKED ISSUE #290: The % operator does not seem to work
Hello!
```python
CONV_LANG = """
def local_sparse_convolution(float(N,C,H,W) I, float(M,KC,KH,KW) W1) -> (O) {{
	O(n, m, h, w) +=! I(n, (c + kc) % c, {sh} * h + kh, {sw} * w + kw) * W1(m, kc, kh, kw)
}}
"""
```
Gives the following error 
`RuntimeError: expected a valid token but found %`

### PR DESCRIPTION
**Pull request by prigoyal**
adding % operator support

closes #290

### CODE REVIEW

File: test/cuda/test_execution_engine.cc [hunk 1]
```diff
@@ -235,6 +235,24 @@ def convolution(float(N,C,H,W) I, float(O,C,KH,KW) W1, float(O) B)
...
+TEST_F(ATenCompilationUnitTest, LocalSparseConvolution) {
+  at::Tensor I = at::CUDA(at::kFloat).rand({N, C, H, W});
+  at::Tensor W1 = at::CUDA(at::kFloat).rand({O, KC, KH, KW});
+  std::vector<at::Tensor> inputs = {I, W1};
+  std::vector<at::Tensor> outputs;
+
+  Check(
+      R"(
+def local_sparse_convolution(float(N, C, H, W) I, float(O, KC, KH, KW) W1) -> (O) {
+    O(n, o, h, w) +=! I(n, (c + kc) % c, h + kh, w + kw) * W1(o, kc, kh, kw) where c in 0:C
```
Comment (abadams) [inline test/cuda/test_execution_engine.cc]: Isn't (c + kc) % c  equal to (kc % c) ?

**Comment by abadams** [comment]
This all makes sense to me so far.

**Comment by ftynse** [comment]
When you get to Halide -> isl conversion, there may be a need to differentiate between right hand sides of `%` that appear in index expressions.  If those are compile-time constants, we can express them as quasi-affine expressions and have exact polyhedral analyses.  Otherwise, we need to overapproximate like we do for indirections.

**Comment by prigoyal** [comment]
thanks @ftynse . I looked further and tried to understand the `tc2halide` and `halide2isl` better. It looks like the segfault happens because of infinite recursion in https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/core/halide2isl.cc#L203 . I fixed that bug and test seems to pass now :)

File: tc/core/halide2isl.cc [hunk 2]
```diff
@@ -35,13 +35,15 @@ using namespace tc::polyhedral::detail;
...
-  // Collect and categorize all the Variable symbols
+  // Collect and categorize all the Halide Variable symbols as reduction
+  // or index variables
   class BuildSymbolTable : public IRVisitor {
     using IRVisitor::visit;
     std::set<std::string> included;
     void visit(const Variable* op) {
       if (!included.count(op->name)) {
         if (op->param.defined()) {
+          // why is this empty? - what's the point of this?
```
Comment (prigoyal) [inline tc/core/halide2isl.cc]: I didn't quite understand why this is empty and added a comment. if one of the reviewers can explain it, that'd be great and I can remove this comment as well

File: tc/core/halide2isl.cc [hunk 3]
```diff
@@ -197,10 +199,9 @@ std::vector<isl::aff> makeIslAffBoundsFromExpr(
   } else if (const Div* op = e.as<Div>()) {
     return combineSingleAffs(space, op, &isl::aff::div);
   } else if (const Mod* op = e.as<Mod>()) {
-    std::vector<isl::aff> result;
     // We cannot span multiple constraints if a modulo operation is involved.
     // x > max(a,b) % C is not equivalent to (x > a % C && x > b % C).
-    auto lhs = makeIslAffBoundsFromExpr(space, e, false, false);
+    auto lhs = makeIslAffBoundsFromExpr(space, op->a, false, false);
```
Comment (prigoyal) [inline tc/core/halide2isl.cc]: using `e` would lead to infinite recursion and hence segfault

File: tc/core/tc2halide.cc [hunk 4]
```diff
@@ -62,43 +62,45 @@ Type translateScalarType(int tcType) {
...
-        Parameter p;
-        if (it != params->end()) {
-          p = it->second;
-        } else {
-          p = Parameter(Int(32), false, 0, d.name(), true);
-          (*params)[d.name()] = p;
-        }
-        dims.push_back(Variable::make(Int(32), p.name(), p));
+  }
+  lang::TensorType type = p.tensorType();
```
Comment (prigoyal) [inline tc/core/tc2halide.cc]: there is no need for `else` here since we return from the `if` condition if satisfied.

File: tc/core/tc2halide.cc [hunk 5]
```diff
@@ -492,20 +496,25 @@ Expr reductionUpdate(Expr e) {
   return Call::make(e.type(), kReductionUpdate, {e}, Call::Intrinsic);
 }
 
+// translate a single TC comprehension/statement to Halide component.
+// funcs, bounds, reductions will be populated
 void translateComprehension(
-    const lang::Comprehension& c,
+    const lang::Comprehension& comprehension,
```
Comment (prigoyal) [inline tc/core/tc2halide.cc]: while reading the code, it was confusing to me to read `c` so I made it more descriptive to facilitate code readability. Hope this is okay, otherwise I can revert too.

File: tc/lang/sema.h [hunk 6]
```diff
@@ -188,23 +191,27 @@ struct Sema {
     }
     return matched_type;
   }
+
```
Comment (prigoyal) [inline tc/lang/sema.h]: I was repeatedly getting confused about where a function started and ended. So I added these white lines to help. is this okay @zdevito ? :)

File: tc/lang/sema.h [hunk 7]
```diff
@@ -366,21 +395,27 @@ struct Sema {
...
+    return createCompound(TK_INT32, anchor->range(), {});
   }
+
   TreeRef dimType(TreeRef anchor) {
     return indexType(anchor);
   }
+
   TreeRef floatType(TreeRef anchor) {
-    return c(TK_FLOAT, anchor->range(), {});
+    return createCompound(TK_FLOAT, anchor->range(), {});
```
Comment (prigoyal) [inline tc/lang/sema.h]: simple `c` was confusing to me as a function name.

File: test/cuda/test_execution_engine.cc [hunk 8]
```diff
@@ -235,6 +235,24 @@ def convolution(float(N,C,H,W) I, float(O,C,KH,KW) W1, float(O) B)
...
+TEST_F(ATenCompilationUnitTest, LocalSparseConvolution) {
+  at::Tensor I = at::CUDA(at::kFloat).rand({N, C, H, W});
+  at::Tensor W1 = at::CUDA(at::kFloat).rand({O, KC, KH, KW});
+  std::vector<at::Tensor> inputs = {I, W1};
+  std::vector<at::Tensor> outputs;
+
+  Check(
+      R"(
+def local_sparse_convolution(float(N, C, H, W) I, float(O, KC, KH, KW) W1) -> (O1) {
+    O1(n, o, h, w) +=! I(n, kc % c, h + kh, w + kw) * W1(o, kc, kh, kw) where c in 0:C
```
Comment (nicolasvasilache) [inline test/cuda/test_execution_engine.cc]: This cannot be correct, it will introduce ```kc % 0``` operations.

File: test/cuda/test_execution_engine.cc [hunk 9]
```diff
@@ -235,6 +235,24 @@ def convolution(float(N,C,H,W) I, float(O,C,KH,KW) W1, float(O) B)
...
+}
+    )",
+      "local_sparse_convolution",
+      tc::CudaMappingOptions::makeConvolutionCudaMappingOptions(),
+      inputs,
+      outputs);
+}
```
Comment (nicolasvasilache) [inline test/cuda/test_execution_engine.cc]: Please add an actual check function (even if you have to write it using plain C loops).

File: tc/lang/sema.h [hunk 10]
```diff
@@ -627,10 +677,12 @@ struct Sema {
     }
     return it == the_env.end() ? nullptr : it->second;
   }
-  TreeRef c(int kind, const SourceRange& range, TreeList&& trees) {
+
+  TreeRef createCompound(int kind, const SourceRange& range, TreeList&& trees) {
```
Comment (nicolasvasilache) [inline tc/lang/sema.h]: We have been using ```makeXYZ``` syntax throughout the codebase for functions that return an owning pointer, let's try to be consistent.

File: tc/lang/sema.h [hunk 11]
```diff
@@ -627,10 +677,12 @@ struct Sema {
...
     return Compound::create(kind, range, std::move(trees));
   }
-  TreeRef s(const std::string& s) {
+
+  TreeRef createString(const std::string& s) {
```
Comment (nicolasvasilache) [inline tc/lang/sema.h]: We have been using ```makeXYZ``` syntax throughout the codebase for functions that return an owning pointer, let's try to be consistent.

File: tc/lang/sema.h (hunk 10 above)
Comment (prigoyal) [reply tc/lang/sema.h]: it was named `create...` because that's how the functions are named in lexer/sema/parser codebade to create a new node like: https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/lang/tree_views.h#L204. The coding conventions recommend to have consistency locally within lexer/sema/parser (for example) code base here https://github.com/facebookresearch/TensorComprehensions/blob/master/CodingConventions.md#naming

if the naming should be `make...`, 1. can you kindly create a rule in `CodingConventions.md` so that people can follow that? 2. it would be also nice to define when should global/local rules be followed? thanks :)

File: tc/lang/sema.h (hunk 11 above)
Comment (prigoyal) [reply tc/lang/sema.h]: [copy-paste from above]:

it was named `create...` because that's how the functions are named in lexer/sema/parser codebade to create a new node like: https://github.com/facebookresearch/TensorComprehensions/blob/master/tc/lang/tree_views.h#L204. The coding conventions recommend to have consistency locally within lexer/sema/parser (for example) code base here https://github.com/facebookresearch/TensorComprehensions/blob/master/CodingConventions.md#naming

if the naming should be `make...`, 1. can you kindly create a rule in `CodingConventions.md` so that people can follow that? 2. it would be also nice to define when should global/local rules be followed? thanks :)

File: tc/lang/sema.h (hunk 10 above)
Comment (nicolasvasilache) [reply tc/lang/sema.h]: sounds good, local trump global :)

File: tc/lang/sema.h (hunk 11 above)
Comment (nicolasvasilache) [reply tc/lang/sema.h]: sounds good, local trumps global :)

File: tc/core/tc2halide.cc [hunk 12]
```diff
@@ -71,37 +71,36 @@ void translateParam(
...
+  vector<Expr> dims;
+  for (auto d_ : type.dims()) {
+    if (d_->kind() == lang::TK_IDENT) {
+      auto d = lang::Ident(d_);
+      auto it = params->find(d.name());
+      Parameter p;
+      if (it != params->end()) {
+        p = it->second;
       } else {
-        CHECK(d_->kind() == lang::TK_CONST);
```
Comment (nicolasvasilache) [inline tc/core/tc2halide.cc]: This hoisting looks like a behavior change to me, could you please explain why it is necessary or correlated with this commit message?

Comment (prigoyal) [reply tc/core/tc2halide.cc]: thanks for the review. I am not sure I understand. There is no code change to this section except removing unnecessary else after line 74. 

line 88 - 98 in the green section matches line 85 - 95 in red section. I think that github alignment for code changes is not exactly good and hence causing confusion.

File: tc/core/halide2isl.cc [hunk 13]
```diff
@@ -93,7 +94,7 @@ namespace {
  * Convert Halide binary expression "op" into a list of isl affine functions by
  * converting its LHS and RHS into lists of affs and concatenating those lists.
  * This is intended to be used with Min/Max operations in upper/lower bound
- * computations, respectively.  Essentially, this allows for replacements
```
Comment (ftynse) [inline tc/core/halide2isl.cc]: I consistently use double spaces between sentences.  Makes it easy to read longer text fragments typeset in monospace fonts.  You may have noticed in my commit messages, or other comments, including this one.  I intend to keep doing so unless the code guidelines require the opposite.

**Comment by nicolasvasilache** [comment]
So it seems the unit test change requested was addressed by simply deleting the unit test. 
As it stands, this PR adds % to parsing and checks that the operation is indeed performed on an int32 value which is good.

However as far as I see, it does not address #290 which requires support for ***modulus array indexing***. So I would remove the  ```halide -> isl``` part from your commit message until this is addressed and a proper test is added.

Could you elaborate on what happens if you try to compile a TC with a modulus array indexing? I assume the parsing now passes but I would expect some undefined behavior when reaching ISL. 

Does the TC mapper complain in an actionable way, just segfault/crash, produce some wrong code or something else?

File: tc/core/tc2halide.cc (hunk 12 above)
Comment (ftynse) [reply tc/core/tc2halide.cc]: diff computation alogrithm is not perfect; reviewing changes per-commit helps

File: tc/core/tc2halide.cc (hunk 5 above)
Comment (ftynse) [reply tc/core/tc2halide.cc]: +1, one character names are for iteration variables of short loops

File: tc/core/tc2halide.cc (hunk 12 above)
Comment (nicolasvasilache) [reply tc/core/tc2halide.cc]: Ah yes, you're right, I see there are 2 columns of numbers. I learned something today! Sorry for the noise here

File: tc/lang/sema.h [hunk 14]
```diff
@@ -447,11 +489,13 @@ struct Sema {
       insert(index_env, index, typ, true);
     }
 
-    // make dimension variables for each dimension of the output tensor
+    // check that the input is not used for output - input are immutable
```
Comment (ftynse) [inline tc/lang/sema.h]: "inputs are" or "input is"

**Comment by ftynse** [comment]
After a brief discussion, it looks like we all want three distinct test cases:
- modulus with constant RHS used in the subscript (`A(i) = B(i % 42)`), make sure _exact_ access relations are computed;
- modulus with variable RHS used in the subscript (`A(i) +=! B(i % j) where j in 1:3`), make sure access relations are non-empty;
- modulus with variable RHS used in the outermost expression (`A(i) = B(i) % C(0)`), check that the result is correct.
The difference in only important when modulus appears in the subscript, where we have a special case for constant RHS.  In the outermost expression, we just check that the most general case works.

Bonus points for adding a warning in the two latter cases that the precondtion (`j != 0` and `C(0) != 0` will not be checked in the runtime)

**Comment by prigoyal** [comment]
thanks all for the review. The unit test in this PR so far is meant to address the "correctness" that % works with a simplest example. The breadth of cases wasn't explored in that unit test. Modulus array indexing also works as it was in earlier version of commit but I don't have a way to write correctness when involving convolution type operations. I'll think of a simple test case to demonstrate that and I welcome any ideas there.

thanks @ftynse for a clear description :) I'll add those unit tests as well.

**Comment by nicolasvasilache** [comment]
> correctness when involving convolution type operations
@prigoyal you could have a simple nested loop that operates on the CPU. It is probably overkill for this PR though our friend in #290 may use it.

> I'll think of a simple test case to demonstrate that and I welcome any ideas there.
You could for instance try an integer array rotation. i.e. [1, 2, 3] -> [3, 1, 2] that should be enough to exercise the ```A(i) = B(i % 42)```, you already have the ```A(i) = B(i) % C(i)``` case. For the ```A(i) +=! B(i % j) where j in 1:3``` case it's a bit more tricky, what @ftynse  suggested (test non-emptyness of access relations) sounds reasonable.

**Comment by ftynse** [comment]
For the first two cases, you could do something similar to what I did for non-affine subscripts [here](https://github.com/facebookresearch/TensorComprehensions/pull/378/files#diff-30ac823bfce2e27b8c04e70f8dba0047R717).  That lets you run do the TC->Halide->Scop conversion with all the parsing and inference involved, but does not require anything runnable.  Given a Scop, you can look at the read/write relations (public member variables).  In the first case, I expect the range of the read relation to have a stride.  It is possible to extract the stride of the isl set using an appropriate isl function (`isl_set_get_stride`). In the second case with the snippet I gave, I expect the range of the read relation to be either a universe or restricted to the extent of the tensor `B`.

For the third case, if that can be expressed as an ATen expression, just have it in `test/cuda/test_tc_mapper`.  Tests do not have to correspond to actual use cases.

File: tc/core/halide2isl.cc (hunk 13 above)
Comment (skimo-openhub) [reply tc/core/halide2isl.cc]: Me too.  In any case, this change does not belong to a commit that adds documentation.

File: tc/core/tc2halide.cc [hunk 15]
```diff
@@ -62,10 +62,13 @@ Type translateScalarType(int tcType) {
   }
 }
 
+// translate the TC def input params to corresponding Halide components.
+// params, inputs will be populated here
```
Comment (skimo-openhub) [inline tc/core/tc2halide.cc]: Please start sentences with a capital and close them with a period.
This also applies to other comments added below.

File: tc/core/tc2halide.cc [hunk 16]
```diff
@@ -62,10 +62,13 @@ Type translateScalarType(int tcType) {
...
 void translateParam(
     const lang::Param& p,
     map<string, Parameter>* params,
     vector<ImageParam>* inputs) {
+  // check if the param is already converted to halide components
```
Comment (skimo-openhub) [inline tc/core/tc2halide.cc]: Hmm... do you mean "has already been converted"?

File: tc/core/tc2halide.cc [hunk 17]
```diff
@@ -492,6 +495,8 @@ Expr reductionUpdate(Expr e) {
   return Call::make(e.type(), kReductionUpdate, {e}, Call::Intrinsic);
 }
 
+// translate a single TC comprehension/statement to Halide component.
```
Comment (skimo-openhub) [inline tc/core/tc2halide.cc]: _a_ Halide component.

File: tc/core/tc2halide.cc [hunk 18]
```diff
@@ -492,6 +495,8 @@ Expr reductionUpdate(Expr e) {
...
+// funcs, bounds, reductions will be populated
```
Comment (skimo-openhub) [inline tc/core/tc2halide.cc]: Since you are documenting this function, you might as well explain what
those are and how they relate to the "Halide component".

File: tc/lang/sema.h [hunk 19]
```diff
@@ -339,6 +339,21 @@ struct Sema {
...
+  // sure that the tree is sematically correct. For example: a variable
+  // may not be input two times. Parser only verifies for the syntax but does
+  // not check the semantics.
+  //
+  // It converts the TK_APPLY nodes to TK_ACCESS or TK_BUILT_IN
+  //
+  // The reduction variables are deduced and the objects are created for them
+  // and they are appended to the tree
+  //
+  // Type checking is also done but by a very small codebase
```
Comment (skimo-openhub) [inline tc/lang/sema.h]: What does "but by a very small codebase" mean?
Does it mean that the type checking is done using a small amount of code?
If so, then why the "but"?

File: tc/lang/sema.h [hunk 20]
```diff
@@ -339,6 +339,21 @@ struct Sema {
...
+  //
+  // withType - associates the type with a given node
```
Comment (skimo-openhub) [inline tc/lang/sema.h]: What's the connection with `checkFunction`?
That is, why did you add this here?

File: tc/lang/sema.h (hunk 11 above)
Comment (skimo-openhub) [reply tc/lang/sema.h]: `createString` doesn't appear to be used anywhere.
It may be better to simply remove it (in a separate commit).

File: tc/core/halide2isl.cc (hunk 3 above)
Comment (skimo-openhub) [reply tc/core/halide2isl.cc]: This comment should be in the commit message.

File: test/cuda/test_corner_cases.cc [hunk 21]
```diff
@@ -285,6 +285,17 @@ TEST(TestCornerCases, E23) {
...
 
+TEST(TestCornerCases, E24) {
+  auto a = I();
+  auto b = I();
+  auto r = I(0);
+  Succeed(
+      "def f(int32 a, int32 b) -> (c) { c(i) = int32(a % b) where i in 0:1 }",
+      {a, b},
+      {r});
+  CHECK_EQ(at::Scalar(r[0]).toInt(), 0);
```
Comment (skimo-openhub) [inline test/cuda/test_corner_cases.cc]: I understand that this file is only meant to check compilation, but shouldn't also add a test
somewhere that checks that the correct operation is being performed?

File: tc/lang/sema.h (hunk 20 above)
Comment (prigoyal) [reply tc/lang/sema.h]: checkFunction is the main entry point for Sema codebase called by `tc2halide`

I added `withType` because this is important component of sema where each node gets it's proper type. I wanted to highlight that it's an important functionality as part of what `checkFunction` will result it.

File: test/cuda/test_corner_cases.cc (hunk 21 above)
Comment (prigoyal) [reply test/cuda/test_corner_cases.cc]: this is the correctness check as well. I can modify it a bit so it actually is visible as correctness test. Currently, I did mental calculation on inputs and added the expected output as 0 but that's not the safe way, I'll change this :)

File: tc/lang/sema.h (hunk 19 above)
Comment (prigoyal) [reply tc/lang/sema.h]: yep, there's no need for "but". I'll remove it :)

File: test/test_cuda_mapper.cc [hunk 22]
```diff
@@ -753,6 +753,26 @@ def perforatedConvolution(float(N, C, H, W) input, float(M, C, KH, KW) weights,
   Prepare(tc);
 }
 
+TEST_F(PolyhedralMapperTest, ModulusConstantRHS) {
+  string tc = R"TC(
+def fun(float(N) a) -> (b) { b(i) = a(i % 2) where i in 0:N }
+)TC";
+  // This triggers tc2halide conversion and should not throw.
+  auto scop = Prepare(tc);
+  auto reads = scop->reads;
```
Comment (prigoyal) [inline test/test_cuda_mapper.cc]: hey @ftynse , I have extracted out the reads here but I don't know how to get the isl stride information. I tried to search in the codebase but couldn't find much help. Can you give some pointers? thanks.

Comment (ftynse) [reply test/test_cuda_mapper.cc]: Yeah, you probably want to search in the third-party/islpp as well :)  There is a greppable documentation file in `doc/`.

Something along the lines of
```c++
for (auto r : scop->reads.range()) {
  if (r.get_tuple_name() != std::string("B")) { // skip irrelevant reads, if any
    continue;
  }
  EXPECT_EQ(r.get_stride(), putExpectedStrideValueHere);
  // EXPECT_TRUE(r.plain_is_universe()); // for the the set being universe
}
```
should work

**Comment by prigoyal** [comment]
the % operator existed in halide2isl but it was segfaulting. 52989ea69bdb93fca1813e5d885a8d9977637eda commit fixed that :)

File: test/test_cuda_mapper.cc (hunk 22 above)
Comment (prigoyal) [reply test/test_cuda_mapper.cc]: thanks. I tried the above snippet but getting compilation failure on `for (auto r : scop->reads.range())` it seems like it's not iterable. 

is the https://github.com/simbuerg/isl/blob/master/doc/user.pod correct file to consult?

Comment (ftynse) [reply test/test_cuda_mapper.cc]: Sorry, my bad, try `scop->reads.range().get_set_list()` instead.

This one https://github.com/nicolasvasilache/isl/blob/ntv_dev/doc/user.pod comes from what we submodule (our copy of isl is ahead even of isl's master)

**Comment by skimo-openhub** [comment]
On Wed, May 02, 2018 at 09:07:31AM -0700, Priya Goyal wrote:
> prigoyal commented on this pull request.
> > @@ -339,6 +339,21 @@ struct Sema {
>          throw ErrorReport(exp) << "NYI - semantic checking for " << exp;
>      }
>    }
> +  // This is the entry function for semantic analysis. It is called by
> +  // tc2halide to associate type with each node of the tree and to also make
> +  // sure that the tree is sematically correct. For example: a variable
> +  // may not be input two times. Parser only verifies for the syntax but does
> +  // not check the semantics.
> +  //
> +  // It converts the TK_APPLY nodes to TK_ACCESS or TK_BUILT_IN
> +  //
> +  // The reduction variables are deduced and the objects are created for them
> +  // and they are appended to the tree
> +  //
> +  // Type checking is also done but by a very small codebase
> +  //
> +  // withType - associates the type with a given node
> 
> checkFunction is the main entry point for Sema codebase called by `tc2halide`
> 
> I added `withType` because this is important component of sema where each node gets it's proper type. I wanted to highlight that it's an important functionality as part of what `checkFunction` will result it.

Hmm... now it just says "Associates the type with a given node".
Is this what checkFunction does or what withType does?
You added that to the documentation of checkFunction.

If you want to refer to some method from inside the documenation
of another method, then please be a bit more explicit about it.
"The method withType can be used to ..."
Just reading this "withType -", I really had no idea what it was about.

skimo

File: test/test_cuda_mapper.cc [hunk 23]
```diff
@@ -753,6 +753,38 @@ def perforatedConvolution(float(N, C, H, W) input, float(M, C, KH, KW) weights,
...
+def fun(float(N) a) -> (b) { b(i) = a(i % 2) where i in 0:N }
+)TC";
+  // This triggers tc2halide conversion and should not throw.
+  auto scop = Prepare(tc);
+  for (auto r : scop->reads.range().get_set_list()) {
+    // skip irrelevant reads, if any
+    if (r.get_tuple_name() != std::string("a")) {
+      continue;
+    }
+    // EXPECT_EQ(r.get_stride(?), ?);
```
Comment (prigoyal) [inline test/test_cuda_mapper.cc]: I am not quite sure what should be the `pos` for `get_stride()`.

Comment (ftynse) [reply test/test_cuda_mapper.cc]: index of the dimension in the set (position of the subscript that should involve the stride), `0` in this case a believe

File: test/test_cuda_mapper.cc [hunk 24]
```diff
@@ -753,6 +753,38 @@ def perforatedConvolution(float(N, C, H, W) input, float(M, C, KH, KW) weights,
...
+def fun(float(N) a) -> (b) { b(i) = a(i % 3) where i in 0:N }
+)TC";
+  // This triggers tc2halide conversion and should not throw.
+  auto scop = Prepare(tc);
+  for (auto r : scop->reads.range().get_set_list()) {
+    // skip irrelevant reads, if any
+    if (r.get_tuple_name() != std::string("a")) {
+      continue;
+    }
+    // EXPECT_EQ(r.get_stride(0), ?);
```
Comment (ftynse) [inline test/test_cuda_mapper.cc]: this test is still commented out...

**Comment by skimo-openhub** [comment]
On Fri, May 04, 2018 at 10:22:09PM +0000, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -753,6 +753,38 @@ def perforatedConvolution(float(N, C, H, W) input, float(M, C, KH, KW) weights,
>    Prepare(tc);
>  }
>  
> +TEST_F(PolyhedralMapperTest, ModulusConstantRHS) {
> +  string tc = R"TC(
> +def fun(float(N) a) -> (b) { b(i) = a(i % 3) where i in 0:N }
> +)TC";
> +  // This triggers tc2halide conversion and should not throw.
> +  auto scop = Prepare(tc);
> +  for (auto r : scop->reads.range().get_set_list()) {
> 
> Sorry, I misled you here; taking the range of the access relation here will actually drop the stride because the stride is on the input dimension.  Instead, let's use _wrapping_.  It is a transformation that reshapes the space from a relation (`[i]->a[a0]`) to a set _wrapping a relation_ (`[[i]->a[a0]]`, note the extra surrounding `[]` that mean it's a set now).  In that set, we can take the stride an compare it with the expected value.  However, we cannot directly check the ex-range's name, so we need to unwrap the set and take its range before using `get_tuple_name()`.

Why do you see this as a stride check?
Why don't you simply compare the access relation to the expected
access relation (perhaps after projecting out the reference tag)?

skimo

File: test/test_cuda_mapper.cc [hunk 25]
```diff
@@ -753,6 +753,38 @@ def perforatedConvolution(float(N, C, H, W) input, float(M, C, KH, KW) weights,
   Prepare(tc);
 }
 
+TEST_F(PolyhedralMapperTest, ModulusConstantRHS) {
+  string tc = R"TC(
+def fun(float(N) a) -> (b) { b(i) = a(i % 3) where i in 0:N }
+)TC";
+  // This triggers tc2halide conversion and should not throw.
+  auto scop = Prepare(tc);
+  for (auto r : scop->reads.range().get_set_list()) {
```
[the remark this answers is not in the document]
Comment (prigoyal) [reply test/test_cuda_mapper.cc]: thanks @ftynse . can you provide guidance on skimo's comment ? I am not sure myself which is the right approach. Thanks a lot :)

[the remark this answers is not in the document]
Comment (ftynse) [reply test/test_cuda_mapper.cc]: It's a matter of preference, I think.  @skimo-openhub suggests to construct a new relation from its textual description, basically `isl::map(ctx, "<paste here the textual representation of the map litearlly>")` (taking `ctx` from whatever isl object available around) and comparing that to the result.  It would work, but I prefer testing only the relevant properties of the relation, hence the stride check.  Checking only relevant properties removes the need to modify the test when irrelevant properties change.  I will accept both ways of checking.

More specifically, current access relations do not involve bounds imposed by tensor sizes.  I'm quite annoyed at this during memory promotion and was planning to add those bounds in access relation construction.  That would change these relations, requiring to update this test that is affected only indirectly.

**Comment by skimo-openhub** [comment]
On Tue, May 08, 2018 at 01:30:16PM -0700, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -753,6 +753,38 @@ def perforatedConvolution(float(N, C, H, W) input, float(M, C, KH, KW) weights,
>    Prepare(tc);
>  }
>  
> +TEST_F(PolyhedralMapperTest, ModulusConstantRHS) {
> +  string tc = R"TC(
> +def fun(float(N) a) -> (b) { b(i) = a(i % 3) where i in 0:N }
> +)TC";
> +  // This triggers tc2halide conversion and should not throw.
> +  auto scop = Prepare(tc);
> +  for (auto r : scop->reads.range().get_set_list()) {
> 
> It's a matter of preference, I think.  @skimo-openhub suggests to construct a new relation from its textual description, basically `isl::map(ctx, "<paste here the textual representation of the map litearlly>")` (taking `ctx` from whatever isl object available around) and comparing that to the result.  It would work, but I prefer testing only the relevant properties of the relation, hence the stride check.  Checking only relevant properties removes the need to modify the test when irrelevant properties change.  I will accept both ways of checking.

Of course you should only check for relevant properties,
but the stride is only a part of the property you want to check.
In any case, don't do any wrapping.  We can easily
add a isl_map_get_range_stride based on isl_map_get_range_stride_info
in the same way isl_set_get_stride is based on
isl_set_get_stride_info and then use isl_map_get_range_stride.

> More specifically, current access relations do not involve bounds imposed by tensor sizes.  I'm quite annoyed at this during memory promotion and was planning to add those bounds in access relation construction.  That would change these relations, requiring to update this test that is affected only indirectly.

You shouldn't check that is _equal_ to some map,
but a _subset_ of the expected property
(and perhaps also that it is non-empty).

skimo

**Comment by ftynse** [comment]
rebased and revived
